apiserver/facades/agent/provisioner/provisioner_test.go

```diff
@@ -1714,3 +1718,109 @@ func (s *withoutControllerSuite) TestMarkMachinesForRemoval(c *gc.C) {
 // 3) Provider could allocate DHCP based addresses on the host device, which would let us
 //    use a bridge on the device and DHCP. (Currently not supported, but desirable for
 //    vSphere and Manual and probably LXD providers.)
+// Addition (manadart 2018-10-09): To begin accommodating the deficiencies noted
+// above, the new suite below uses mocks for tests ill-suited to the dummy
+// provider. We could reasonably re-write the tests above over time to use the
+// new suite.
+
+type provisionerMockSuite struct {
+	coretesting.BaseSuite
+
+	environ      *environtesting.MockNetworkingEnviron
+	host         *MockMachine
+	container    *MockMachine
+	device       *MockLinkLayerDevice
+	parentDevice *MockLinkLayerDevice
+}
+
+var _ = gc.Suite(&provisionerMockSuite{})
+
+// Even when the provider supports container addresses, manually provisioned
+// machines should fall back to DHCP.
+func (s *provisionerMockSuite) TestManuallyProvisionedHostsUseDHCPForContainers(c *gc.C) {
+	defer s.setup(c).Finish()
+
+	s.expectManuallyProvisionedHostsUseDHCPForContainers()
+
+	res := params.MachineNetworkConfigResults{
+		Results: []params.MachineNetworkConfigResult{{}},
+	}
+	ctx := provisioner.NewPrepareOrGetContext(res, false)
+
+	// ProviderCallContext is not required by this logical path; we pass nil.
+	err := ctx.ProcessOneContainer(s.environ, nil, 0, s.host, s.container)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(res.Results[0].Config, gc.HasLen, 1)
+
+	cfg := res.Results[0].Config[0]
+	c.Check(cfg.ConfigType, gc.Equals, "dhcp")
+	c.Check(cfg.ProviderSubnetId, gc.Equals, "")
+	c.Check(cfg.VLANTag, gc.Equals, 0)
+}
+
+func (s *provisionerMockSuite) expectManuallyProvisionedHostsUseDHCPForContainers() {
```

Comment:
This feels big enough that it might  be good to break it down into several isolation functions that each define one bit of behavior. We can leave it as a top level expectation, but it takes quite a bit to parse what are the interactions between the various stages. And not all of them are the "interesting" critical part.